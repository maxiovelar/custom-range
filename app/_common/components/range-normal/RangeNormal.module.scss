.wrapper {
    display: flex;
    gap: 30px;
}

.value {
    border-radius: 8px;
}

.container {
    position: relative;
    display: flex;
    justify-content: center;
    align-items: center;
    width: 24rem;
    height: 1.25rem;
}

.range {
    position: absolute;
    background-color: gray;
    display: flex;
    align-items: center;
    justify-content: center;
    width: 24rem;
    height: 0.75rem;
    border-radius: 1.5rem;
}

.min,
.max {
    position: absolute;
    z-index: 2;
    width: 1.75rem;
    height: 1.75rem;
    border-radius: 50%;
    background-color: #e13d3d;

    &:hover {
        width: 2.5rem;
        height: 2.5rem;
    }
}

.min {
    left: 0;
}

.max {
    right: 0;
}

.dragging {
    background-color: green !important;
}