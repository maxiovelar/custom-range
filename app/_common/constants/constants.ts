import { type PageUrl, type Link, type RangeType } from "@/_types/types"

export const links: Record<string, Link> = {
    home: {
        href: "/",
        text: "Home",
    },
    menu1: {
        href: "/exercise1",
        text: "Exercise 1",
    },
    menu2: {
        href: "/exercise2",
        text: "Exercise 2",
    },
    github: {
        href: "https://github.com/maxiovelar",
        text: "Maximiliano Ovelar",
    },
    nextjs: {
        href: "https://nextjs.org/",
        text: "Next.js",
    },
};

export const LITERALS = {
    minValue: 'min-value',
    maxValue: 'max-value',
    minSelector: 'min-selector',
    maxSelector: 'max-selector',
    range: 'range',
    progress: 'progress',
    currency: 'â‚¬',
};

export const pageTitles: Record<PageUrl, string> = {
    home: "Custom Ranges Exercise",
    exercise1: "Normal Range",
    exercise2: "Fixed Values Range",
};

export const paths: Record<RangeType, string> = {
    rangeNormal: "ca276e65-744d-4dc8-82a1-ee073aaf987b",
    fixedValues: "15e130c5-a7bb-40ec-906a-c7f1d1962e39",
}

export const rangeFixedConstants = {
    selectorWidth: 28,
    rangeWidth: 512,
}
