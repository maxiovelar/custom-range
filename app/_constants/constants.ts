import { type PageUrl, type Link } from "@/_types/types"

export const links: Record<string, Link> = {
    home: {
        href: "/",
        text: "Home",
    },
    menu1: {
        href: "/exercise1",
        text: "Exercise 1",
    },
    menu2: {
        href: "/exercise2",
        text: "Exercise 2",
    },
    github: {
        href: "https://github.com/maxiovelar",
        text: "Maximiliano Ovelar",
    },
    nextjs: {
        href: "https://nextjs.org/",
        text: "Next.js",
    },
};

export const pageTitles: Record<PageUrl, string> = {
    home: "Custom Ranges Exercise",
    exercise1: "Normal Range",
    exercise2: "Fixed Values Range",
};