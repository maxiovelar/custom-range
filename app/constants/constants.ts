import { type PageUrl, type Link, type RangeType } from "@/types/types"

export const links: Record<string, Link> = {
    home: {
        href: "/",
        text: "Home",
    },
    menu1: {
        href: "/exercise1",
        text: "Exercise 1",
    },
    menu2: {
        href: "/exercise2",
        text: "Exercise 2",
    },
    github: {
        href: "https://github.com/maxiovelar",
        text: "Maximiliano Ovelar",
    },
    nextjs: {
        href: "https://nextjs.org/",
        text: "Next.js",
    },
};

export const pageTitles: Record<PageUrl, string> = {
    home: "Custom Ranges Exercise",
    exercise1: "Normal Range",
    exercise2: "Fixed Values Range",
};

export const endpoints: Record<RangeType, string> = {
    normalRange: "ca276e65-744d-4dc8-82a1-ee073aaf987b",
    fixedValues: "948df8dc-397c-4e6c-a7f6-f24f8b54fb7e",
}