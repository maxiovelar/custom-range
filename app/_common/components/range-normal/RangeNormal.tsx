'use client';

import React, { useEffect, useState } from 'react';
import styles from './RangeNormal.module.scss';
import {
    onChangeMaxValueHandler,
    onChangeMinValueHandler,
    hasText,
    selectorMoveHandler,
} from '@/_infra/services/mouseEventService';
import { LITERALS } from '@/_common/constants/constants';
import { setState } from '@/store';

export interface RangeNormalProps {
    min: number;
    max: number;
}

export const RangeNormal = ({ min, max }: RangeNormalProps) => {
    const [inputMinValue, setInputMinValue] = useState(min);
    const [inputMaxValue, setInputMaxValue] = useState(max);

    useEffect(() => {
        const minSelector = document.getElementById(LITERALS.minSelector) as HTMLElement;
        const maxSelector = document.getElementById(LITERALS.maxSelector) as HTMLElement;
        const range = document.getElementById(LITERALS.range) as HTMLElement;
        const rangeWidth = range.offsetWidth - maxSelector.offsetWidth;
        const rangeStartPosition = range.getBoundingClientRect().left;

        setState({ type: 'SET_MIN', payload: min });
        setState({ type: 'SET_MAX', payload: max });
        setState({ type: 'SET_RANGE_START_POSITION', payload: rangeStartPosition });
        setState({ type: 'SET_RANGE_END_POSITION', payload: rangeStartPosition + rangeWidth });

        selectorMoveHandler(minSelector, setInputMinValue);
        selectorMoveHandler(maxSelector, setInputMaxValue);
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    const [error, setError] = useState('');

    const handleChangeMinValue = (event: React.ChangeEvent<HTMLInputElement>) => {
        const newValue = Number(event.target.value);
        setInputMinValue(newValue);
        onChangeMinValueHandler(newValue, setError);
    };

    const handleChangeMaxValue = (event: React.ChangeEvent<HTMLInputElement>) => {
        const newValue = Number(event.target.value);
        setInputMaxValue(newValue);
        onChangeMaxValueHandler(newValue, setError);
    };

    const isError = hasText(error);

    return (
        <div className={styles.wrapper}>
            <div>
                <input
                    type="number"
                    id={LITERALS.minValue}
                    className={styles.value}
                    defaultValue={inputMinValue}
                    onChange={handleChangeMinValue}
                    min={min}
                    max={max}
                />
                <label htmlFor={LITERALS.minValue} className={styles.label}>
                    {LITERALS.currency}
                </label>
            </div>
            <div className={styles.container}>
                <div id={LITERALS.range} className={styles.range}>
                    <div id={LITERALS.progress} className={styles.progress}></div>
                </div>
                <button
                    type="button"
                    id={LITERALS.minSelector}
                    draggable={!isError}
                    className={styles.min}
                    disabled={isError}
                ></button>
                <button
                    type="button"
                    id={LITERALS.maxSelector}
                    draggable={!isError}
                    className={styles.max}
                    disabled={isError}
                ></button>
            </div>
            <div>
                <input
                    type="number"
                    id={LITERALS.maxValue}
                    className={styles.value}
                    defaultValue={inputMaxValue}
                    onChange={handleChangeMaxValue}
                    min={min}
                    max={max}
                />
                <label htmlFor={LITERALS.maxValue} className={styles.label}>
                    {LITERALS.currency}
                </label>
            </div>
            {isError && <span className={styles.error}>{error}</span>}
        </div>
    );
};
