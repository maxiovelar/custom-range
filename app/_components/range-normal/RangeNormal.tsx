'use client';

import React, { useEffect, useState } from 'react';
import styles from './RangeNormal.module.scss';
import {
    onChangeMaxValueHandler,
    onChangeMinValueHandler,
    hasText,
    selectorMoveHandler,
} from '@/_business/rangeNormal';
import { LITERALS } from '@/_constants/constants';
import { setState } from '@/store';

export interface RangeNormalProps {
    min: number;
    max: number;
}

export const RangeNormal = ({ min, max }: RangeNormalProps) => {
    const [inputMinValue, setInputMinValue] = useState<string>(`${min}`);
    const [inputMaxValue, setInputMaxValue] = useState<string>(`${max}`);
    const [error, setError] = useState('');
    const isError = hasText(error);

    useEffect(() => {
        const minSelector = document.getElementById(LITERALS.minSelector) as HTMLElement;
        const maxSelector = document.getElementById(LITERALS.maxSelector) as HTMLElement;
        const range = document.getElementById(LITERALS.range) as HTMLElement;
        const rangeWidth = range.offsetWidth - maxSelector.offsetWidth;
        const rangeStartPosition = range.getBoundingClientRect().left;

        setState({ type: 'SET_MIN', payload: min });
        setState({ type: 'SET_MAX', payload: max });
        setState({ type: 'SET_RANGE_START_POSITION', payload: rangeStartPosition });
        setState({ type: 'SET_RANGE_END_POSITION', payload: rangeStartPosition + rangeWidth });

        selectorMoveHandler(minSelector, setInputMinValue);
        selectorMoveHandler(maxSelector, setInputMaxValue);
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    const handleChangeMinValue = (event: React.ChangeEvent<HTMLInputElement>) => {
        const newValue = event.target.value;
        setInputMinValue(newValue);
        onChangeMinValueHandler(Number(newValue), setError);
    };

    const handleChangeMaxValue = (event: React.ChangeEvent<HTMLInputElement>) => {
        const newValue = event.target.value;
        setInputMaxValue(newValue);
        onChangeMaxValueHandler(Number(newValue), setError);
    };

    return (
        <div className={styles.wrapper}>
            <div>
                <input
                    type="number"
                    id={LITERALS.minValue}
                    aria-label={LITERALS.minValue}
                    className={styles.value}
                    value={inputMinValue}
                    onChange={handleChangeMinValue}
                    min={min}
                    max={max}
                />
                <label htmlFor={LITERALS.minValue} className={styles.label}>
                    {LITERALS.currency}
                </label>
            </div>
            <div className={styles.container}>
                <div id={LITERALS.range} data-testid={LITERALS.range} className={styles.range}>
                    <div id={LITERALS.progress} data-testid={LITERALS.progress} className={styles.progress}></div>
                </div>
                <button
                    type="button"
                    id={LITERALS.minSelector}
                    aria-label={LITERALS.minSelector}
                    draggable={!isError}
                    className={styles.min}
                    disabled={isError}
                ></button>
                <button
                    type="button"
                    id={LITERALS.maxSelector}
                    aria-label={LITERALS.maxSelector}
                    draggable={!isError}
                    className={styles.max}
                    disabled={isError}
                ></button>
            </div>
            <div>
                <input
                    type="number"
                    id={LITERALS.maxValue}
                    aria-label={LITERALS.maxValue}
                    className={styles.value}
                    value={inputMaxValue}
                    onChange={handleChangeMaxValue}
                    min={min}
                    max={max}
                />
                <label htmlFor={LITERALS.maxValue} className={styles.label}>
                    {LITERALS.currency}
                </label>
            </div>
            {isError && <span className={styles.error}>{error}</span>}
        </div>
    );
};
