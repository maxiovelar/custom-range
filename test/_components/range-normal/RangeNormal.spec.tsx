import React from 'react';
import { render, screen } from '@testing-library/react';
import { RangeNormal } from '@/_components/range-normal/RangeNormal';
import { LITERALS } from '@/_constants/constants';

describe('RangeNormal', () => {
    const min = 1;
    const max = 100;

    beforeEach(() => {
        render(<RangeNormal min={min} max={max} />);
    });

    it('renders the component with correct min and max values', () => {
        const minInputValue = screen.getByLabelText('min-value');
        const maxInputValue = screen.getByLabelText('max-value');

        expect(minInputValue).toHaveValue(min);
        expect(maxInputValue).toHaveValue(max);
    });

    it('renders the component with range and progress elements', () => {
        expect(screen.getByTestId(LITERALS.range)).toBeInTheDocument();
        expect(screen.getByTestId(LITERALS.progress)).toBeInTheDocument();
    });
});
