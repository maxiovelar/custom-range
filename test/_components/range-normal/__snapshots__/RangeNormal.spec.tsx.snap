// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeNormal should render correctly 1`] = `
<div>
  <div
    class="wrapper"
  >
    <div>
      <input
        aria-label="min-value"
        class="value"
        id="min-value"
        max="100"
        min="1"
        type="number"
        value="1"
      />
      <label
        class="label"
        for="min-value"
      >
        €
      </label>
    </div>
    <div
      class="container"
    >
      <div
        class="range"
        data-testid="range"
        id="range"
      >
        <div
          class="progress"
          id="progress"
        />
      </div>
      <button
        aria-label="min-selector"
        class="min"
        draggable="true"
        id="min-selector"
        type="button"
      />
      <button
        aria-label="max-selector"
        class="max"
        draggable="true"
        id="max-selector"
        type="button"
      />
    </div>
    <div>
      <input
        aria-label="max-value"
        class="value"
        id="max-value"
        max="100"
        min="1"
        type="number"
        value="100"
      />
      <label
        class="label"
        for="max-value"
      >
        €
      </label>
    </div>
  </div>
</div>
`;
